flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBaseTest.java

```diff
@@ -0,0 +1,469 @@
+	@Ignore
+	@Test
+	public void testAsyncErrorThrownOnInvoke() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+		MockCassandraSinkBase casSinkFunc = new MockCassandraSinkBase<>(builder);
+		casSinkFunc.setFlushOnCheckpoint(true);
+
+		casSinkFunc.open(new Configuration());
+
+		casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_FAILURE);
+		try {
+			casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_SUCCESS);
+		} catch (IOException e) {
+			//expected async error thrown from invoke()
+
+			//Final pending updates should be zero
+			Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+
+			casSinkFunc.close();
+
+			//done
+			return;
+		}
+
+		Assert.fail();
+	}
+
+	/**
+	 * Test ensures that an asyncError would be thrown when checkpoint performs if previously message delivery failed.
+	 */
+	@Test
+	public void testAsyncErrorThrownOnCheckpoint() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+		MockCassandraSinkBase casSinkFunc = new MockCassandraSinkBase<>(builder);
+		casSinkFunc.setFlushOnCheckpoint(true);
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness =
+			new OneInputStreamOperatorTestHarness<>(new StreamSink<>(casSinkFunc));
+
+		testHarness.open();
+
+		casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_FAILURE);
+
+		try {
+			testHarness.snapshot(123L, 123L);
+		} catch (Exception e) {
+			//expected async error from snapshotState()
+
+			Assert.assertTrue(e.getCause() instanceof IllegalStateException);
+			Assert.assertTrue(e.getCause().getMessage().contains("Failed to send data to Cassandra"));
+
+			//Final pending updates should be zero
+			Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+
+			casSinkFunc.close();
+
+			//done
+			return;
+		}
+
+		Assert.fail();
+	}
+
+	///////////////////////
+	// Multi Thread Test
+	///////////////////////
+
+	/**
+	 * Test ensures that CassandraSinkBase would flush all in-flight message on close(), accompanied with concurrent
+	 * message delivery successfully via a thraedpool.
+	 */
+	@Test
+	public void testFlushOnPendingRecordsOnCloseWithSuccessfulMessage() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+
+		ExecutorService threadPool = Executors.newFixedThreadPool(MAX_THREAD_POOL_SIZE);
+		MockCassandraSinkBase casSinkFunc = new MockCassandraSinkBase<>(builder);
+		casSinkFunc.setFlushOnCheckpoint(true);
+
+		casSinkFunc.open(new Configuration());
+
+		for (int i = 0; i < MAX_THREAD_NUM; i++) {
+			threadPool.submit(() -> {
+				try {
+
+					casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.DELAYED_SUCCESS);
+				} catch (Exception e) {
+					LOG.error("Error while dispatching sending message to Cassandra sink => {} ", e);
+				}
+			});
+		}
+
+		//wait until the first message has been dispatched and invoked
+		Thread.sleep(500);
```

Comment:
these waits are too long. They increase the test execution times, and are usually a sign that the test isn't really stable.

Reply:
Will rewrite these with more deterministic way, similar from the comment above.